src/ages/distributions/ageThreeDistribution.ts

```diff
@@ -0,0 +1,75 @@
+import { providers } from "ethers";
+import { EpochConfig } from "../ages.types";
+import { AgeDistribution } from "./distributions.types";
+import tokens from "@morpho-labs/morpho-ethers-contract/lib/tokens";
+import { parseUnits } from "ethers/lib/utils";
+import getMarketsData from "../../utils/markets/getMarketsData";
+import fetchProposal from "../../utils/snapshot/fetchProposal";
+import { MarketEmission } from "../../utils";
+
+const MARKETS = {
+  cDAI: tokens.dai.cToken!,
+  cUSDC: tokens.usdc.cToken!,
+  cUSDT: tokens.usdt.cToken!,
+  cWBTC: tokens.wBtc.cToken!,
+  cETH: tokens.wEth.cToken!,
+  cCOMP: tokens.comp.cToken!,
+  cUNI: tokens.uni.cToken!,
+  aDai: tokens.dai.aToken!,
+  aUSDC: tokens.usdc.aToken!,
+  aUSDT: tokens.usdt.aToken!,
+  aWBTC: tokens.wBtc.aToken!,
+  aWETH: tokens.wEth.aToken!,
+  aUNI: tokens.uni.aToken!,
+  aSTETH: tokens.stEth.aToken!,
+};
+
+export const ageThreeDistribution = async (
+  ageConfig: AgeDistribution,
+  epochConfig: EpochConfig,
+  provider?: providers.Provider
+) => {
+  if (!epochConfig.snapshotBlock) throw Error(`Cannot distribute tokens for epoch ${epochConfig.id}: no snapshotBlock`);
+  if (!epochConfig.snapshotProposal)
+    throw Error(`Cannot distribute tokens for epoch ${epochConfig.id}: no snapshotProposal`);
+  const proposal = await fetchProposal(epochConfig.snapshotProposal);
+  if (proposal.state !== "closed")
+    throw Error(
+      `Cannot distribute tokens for epoch ${epochConfig.id}: proposal ${epochConfig.snapshotProposal} is not closed`
+    );
+  const duration = epochConfig.finalTimestamp.sub(epochConfig.initialTimestamp);
+  const { markets } = await getMarketsData(epochConfig.snapshotBlock, provider!);
+  const totalScoreBn = parseUnits(proposal.scores_total);
+  const marketsEmissions = Object.fromEntries(
+    proposal.scores.map((score, index) => {
+      const symbol = proposal.choices[index];
+      const address = MARKETS[symbol as keyof typeof MARKETS];
+      if (!address) throw Error(`Cannot distribute tokens for epoch ${epochConfig.id}: unknown market ${symbol}`);
+      const marketData = markets.find((market) => market.address.toLowerCase() === address.toLowerCase());
+      if (!marketData)
+        throw Error(`Cannot distribute tokens for epoch ${epochConfig.id}: no market data for ${symbol}`);
+      const scoreBn = parseUnits(score.toString());
+      const distribution = epochConfig.totalEmission.mul(scoreBn).div(totalScoreBn);
+      const total = marketData.totalBorrow.add(marketData.totalSupply);
+      const supply = marketData.totalSupply.mul(distribution).div(total);
+      const supplyRate = supply.div(duration);
+      const borrow = marketData.totalBorrow.mul(distribution).div(total);
+      const borrowRate = borrow.div(duration);
```

Comment:
could test that the two rates are equal

Reply:
mm why?

Reply:
they should be equal

Reply:
supplyRate = marketRate * totalSupply / (totalSupply + totalBorrow) 

Why it should be equal to the borrow rate?

Reply:
I thought that the supply/borrow rate was the number of MORPHO earned per second for one unit of underlying token (what you wrote, divided by the totalSupply/totalBorrow). Here that's not the case, so there is no reason why they would be equal you are right (but the naming "rate" is misleading then IMO).